proof/refine/ARM/Ipc_R.thy

```diff
@@ -3156,6 +3156,461 @@ lemma maybeReturnSc_corres:
+                   split: Structures_A.kernel_object.splits)
+    apply (prop_tac "obj_at (same_caps (kernel_object.SchedContext sc n)) ptr a")
+     apply (clarsimp simp: obj_at_def)
+    apply (clarsimp simp: state_relation_def
+                          z[simplified obj_at'_def is_sc_obj_def projectKO_eq projectKO_opts_defs, simplified])
+    apply (clarsimp simp: caps_of_state_after_update cte_wp_at_after_update
+                              swp_def fun_upd_def obj_at_def)
+    apply (subst conj_assoc[symmetric])
+    apply (rule conjI[rotated])
+     apply (clarsimp simp: ghost_relation_def)
+     apply (erule_tac x=ptr in allE)+
+     apply (clarsimp simp: obj_at_def a_type_def
+                     split: Structures_A.kernel_object.splits if_split_asm)
+    apply (fold fun_upd_def)
+    apply (simp only: pspace_relation_def simp_thms
+                      pspace_dom_update[where x="kernel_object.SchedContext _ _"
+                                          and v="kernel_object.SchedContext (f _) _",
+                                        simplified a_type_def, simplified])
+    apply (simp only: dom_fun_upd2 simp_thms)
+    apply (elim conjE)
+    apply (frule bspec, erule domI)
+    apply (rule conjI)
+     apply (rule ballI, drule(1) bspec)
+     apply (drule domD)
+     apply (clarsimp simp: project_inject split: if_split_asm kernel_object.split_asm)
+     apply (rename_tac bb aa ba)
+     apply (drule_tac x="(aa, ba)" in bspec, simp)
+     apply clarsimp
+     apply (frule_tac ko'="kernel_object.SchedContext sc n" and x'=ptr
+              in obj_relation_cut_same_type)
+        apply simp+
+     apply (clarsimp simp: a_type_def split: Structures_A.kernel_object.split_asm if_split_asm)
+      (* sc_replies_relation *)
+    apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def map_project_def scs_of_kh_def)
+    apply (drule_tac x=p in spec)
+    by (auto simp: typ_at'_def ko_wp_at'_def opt_map_def sc_of_def projectKO_opts_defs
+            split: if_splits)
+qed
+
+lemma setSchedContext_stack_update_corres:
+  "corres dc
+         (\<lambda>s. kheap s ptr = Some (kernel_object.SchedContext sc n))
+         (ko_at' sc' ptr and (\<lambda>s'. heap_list (replyPrevs_of s') (F (scReply sc')) (f (sc_replies sc))))
+            (set_object ptr (kernel_object.SchedContext (sc_replies_update f sc) n))
+            (setSchedContext ptr (scReply_update F sc'))"
+  apply (rule corres_guard_imp)
+    apply (rule setSchedContext_sc_replies_relation_corres[where f'="scReply_update F" and f="sc_replies_update f"])
+     apply (clarsimp simp: sc_relation_def)
+    apply (clarsimp simp: objBits_def objBitsKO_def)
+   apply clarsimp+
+  done
+
+(* FIXME RT: move *)
+lemma list_case_when:
+  "(case l of [] \<Rightarrow> return ()
+            | r # x \<Rightarrow> (f r x))
+ = (when (l \<noteq> []) $ f (hd l) (tl l))"
+  by (clarsimp split: list.splits)
+
+lemma heap_list_heap_update_noop:
```

Comment:
Move to Heap_List.thy?

Reply:
This one also exists in my misc lemma PR (though different name), and this one is now merged as part of the sym_heap PR (and it’s in Heap_List.thy :) ).